From 86bfb093a92b050b323ea790934944874e978396 Mon Sep 17 00:00:00 2001
From: Tyrone Ting <kfting@nuvoton.com>
Date: Tue, 19 Aug 2025 14:41:17 +0800
Subject: [PATCH 2/3] drivers: i2c: auto recover when i2c single-master bus is
 busy

To enable multi-master mode, add "multi-master = <1>;" in the i2c node(s)
of the dts or overlay file.

In multi-master mode, the i2c bus doesn't do auto-recovery when
the bus is busy.

For example:

&i2c1a {
       clock-frequency = <I2C_BITRATE_STANDARD>;
       multi-master = <1>;
       status = "okay";
};

Signed-off-by: Tyrone Ting <kfting@nuvoton.com>
---
 drivers/i2c/i2c_npcm4xx.c                 | 16 +++++++++++++---
 dts/bindings/i2c/nuvoton,npcm4xx-i2c.yaml |  5 +++++
 2 files changed, 18 insertions(+), 3 deletions(-)

diff --git a/drivers/i2c/i2c_npcm4xx.c b/drivers/i2c/i2c_npcm4xx.c
index 575dee1d0b55..2ea58a1ba9b5 100755
--- a/drivers/i2c/i2c_npcm4xx.c
+++ b/drivers/i2c/i2c_npcm4xx.c
@@ -76,6 +76,7 @@ struct i2c_npcm4xx_config {
 	uint32_t wait_free_time;
 	uint32_t scllt;
 	uint32_t sclht;
+	uint8_t multi_master;
 };
 
 /*rx_buf and tx_buf address must 4-align for DMA */
@@ -1134,9 +1135,17 @@ static int i2c_npcm4xx_transfer(const struct device *dev, struct i2c_msg *msgs,
 	}
 
 	if (bus_busy) {
-		inst->SMBnCST |= BIT(NPCM4XX_SMBnCST_BB);
-		i2c_npcm4xx_mutex_unlock(dev);
-		return -EAGAIN;
+		if (!config->multi_master) {
+			ret = i2c_npcm4xx_recover_bus(dev);
+			if (ret) {
+				i2c_npcm4xx_mutex_unlock(dev);
+				return ret;
+			}
+		} else {
+			inst->SMBnCST |= BIT(NPCM4XX_SMBnCST_BB);
+			i2c_npcm4xx_mutex_unlock(dev);
+			return -EAGAIN;
+		}
 	}
 
 	/* prepare data to transfer */
@@ -1251,6 +1260,7 @@ static const struct i2c_driver_api i2c_npcm4xx_driver_api = {
 				  0),	 		                         \
 		.sclht = DT_INST_PROP_OR(inst, sclht,		                 \
 				  0),	 		                         \
+		.multi_master = DT_INST_PROP_OR(inst, multi_master, 0),          \
 	};									 \
 										 \
 	static struct i2c_npcm4xx_data i2c_npcm4xx_data_##inst;			 \
diff --git a/dts/bindings/i2c/nuvoton,npcm4xx-i2c.yaml b/dts/bindings/i2c/nuvoton,npcm4xx-i2c.yaml
index ba8faac2edcf..680c08cba8f8 100644
--- a/dts/bindings/i2c/nuvoton,npcm4xx-i2c.yaml
+++ b/dts/bindings/i2c/nuvoton,npcm4xx-i2c.yaml
@@ -26,3 +26,8 @@ properties:
         type: int
         required: false
         description: value from 5 to 255
+    multi-master:
+        type: int
+        required: false
+        description: |
+            i2c multi-master support. Default value is disabled.
-- 
2.17.1

